cpp/src/arrow/util/bit_block_counter.h

```diff
@@ -586,5 +961,144 @@ static void VisitTwoBitBlocksVoid(const std::shared_ptr<Buffer>& left_bitmap_buf
+      return VisitBitBlocksVoid(right_bitmap_buf, right_offset, length,
+                                std::forward<VisitNotNull>(visit_not_null),
+                                std::forward<VisitNull>(visit_null));
+    } else if ((mid_bitmap_buf == NULLPTR) && (right_bitmap_buf == NULLPTR)) {
+      return VisitBitBlocksVoid(left_bitmap_buf, left_offset, length,
+                                std::forward<VisitNotNull>(visit_not_null),
+                                std::forward<VisitNull>(visit_null));
+    } else {
+      return VisitBitBlocksVoid(mid_bitmap_buf, mid_offset, length,
+                                std::forward<VisitNotNull>(visit_not_null),
+                                std::forward<VisitNull>(visit_null));
+    }
+  }
+  // Two bitmaps are present
+  if (left_bitmap_buf == NULLPTR) {
+    const uint8_t* mid_bitmap = mid_bitmap_buf->data();
+    const uint8_t* right_bitmap = right_bitmap_buf->data();
+    BinaryBitBlockCounter bit_counter(mid_bitmap, mid_offset, right_bitmap, right_offset,
+                                      length);
+    int64_t position = 0;
+    while (position < length) {
+      BitBlockCount block = bit_counter.NextAndWord();
+      if (block.AllSet()) {
+        for (int64_t i = 0; i < block.length; ++i, ++position) {
+          visit_not_null(position);
+        }
+      } else if (block.NoneSet()) {
+        for (int64_t i = 0; i < block.length; ++i, ++position) {
+          visit_null();
+        }
+      } else {
+        for (int64_t i = 0; i < block.length; ++i, ++position) {
+          if (bit_util::GetBit(mid_bitmap, mid_offset + position) &&
+              bit_util::GetBit(right_bitmap, right_offset + position)) {
+            visit_not_null(position);
+          } else {
+            visit_null();
+          }
+        }
+      }
+    }
+  } else if (mid_bitmap_buf == NULLPTR) {
+    const uint8_t* left_bitmap = left_bitmap_buf->data();
+    const uint8_t* right_bitmap = right_bitmap_buf->data();
+    BinaryBitBlockCounter bit_counter(left_bitmap, left_offset, right_bitmap,
+                                      right_offset, length);
+    int64_t position = 0;
+    while (position < length) {
+      BitBlockCount block = bit_counter.NextAndWord();
+      if (block.AllSet()) {
+        for (int64_t i = 0; i < block.length; ++i, ++position) {
+          visit_not_null(position);
+        }
+      } else if (block.NoneSet()) {
+        for (int64_t i = 0; i < block.length; ++i, ++position) {
+          visit_null();
+        }
+      } else {
+        for (int64_t i = 0; i < block.length; ++i, ++position) {
+          if (bit_util::GetBit(left_bitmap, left_offset + position) &&
+              bit_util::GetBit(right_bitmap, right_offset + position)) {
+            visit_not_null(position);
+          } else {
+            visit_null();
+          }
+        }
+      }
+    }
+  } else {
+    const uint8_t* left_bitmap = left_bitmap_buf->data();
```

Comment:
Is this clause correct?
It seems that `right_bitmap_buf` may be ignored.

Reply:
Thanks. Better to test each of the cases separately.